File: frontend/src/types/interview.ts
export interface InterviewQuestion {
  question: string;
  category: "company_research" | "technical" | "behavioral" | "culture_fit";
  difficulty: 1 | 2 | 3;
  follow_up_hints: string[];
}

export interface InterviewSetupRequest {
  job_id?: string;
  job_description?: string;
  job_title?: string;
  company_name?: string;
}

export interface InterviewSetupResponse {
  prep_id: string;
  status: "generating" | "completed" | "failed";
}

export interface InterviewPrepStatusResponse {
  prep_id: string;
  status: "generating" | "completed" | "failed";
  question_count: number;
}

export interface InterviewPrepListItem {
  id: string;
  job_id: string | null;
  job_title: string | null;
  company_name: string | null;
  status: string;
  question_count: number;
  created_at: string;
}

export interface InterviewPrepListResponse {
  preps: InterviewPrepListItem[];
  total: number;
}

export interface ChatEvaluation {
  score: number;
  strengths: string[];
  improvements: string[];
  coaching_tip: string;
  sample_answer: string;
}

export interface ChatQuestion {
  question: string;
  category: string;
  difficulty: number;
  follow_up_hints: string[];
}

export interface InterviewChatRequest {
  prep_id: string;
  session_id?: string;
  answer?: string;
}

export interface InterviewChatResponse {
  session_id: string;
  evaluation: ChatEvaluation | null;
  next_question: ChatQuestion | null;
  session_complete: boolean;
  difficulty: number;
  overall_feedback: string | null;
}

export interface SessionMessage {
  role: "question" | "user" | "coach" | "summary";
  content: string;
  score: number | null;
  category: string | null;
  difficulty: number | null;
}

export interface InterviewSessionDetail {
  id: string;
  prep_id: string;
  status: string;
  current_difficulty: number;
  questions_answered: number;
  overall_score: number | null;
  messages: SessionMessage[];
  overall_feedback: string | null;
  completed_at: string | null;
  created_at: string;
}

export interface InterviewSessionListItem {
  id: string;
  prep_id: string;
  job_title: string | null;
  company_name: string | null;
  status: string;
  questions_answered: number;
  overall_score: number | null;
  created_at: string;
}

export interface InterviewSessionListResponse {
  sessions: InterviewSessionListItem[];
  total: number;
}

File: backend/app/schemas/interview.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field, model_validator


class InterviewSetupRequest(BaseModel):
    job_id: uuid.UUID | None = None
    job_description: str | None = Field(default=None, max_length=10000)
    job_title: str | None = Field(default=None, max_length=200)
    company_name: str | None = Field(default=None, max_length=200)

    @model_validator(mode="after")
    def check_at_least_one_source(self) -> "InterviewSetupRequest":
        if self.job_id is None and not self.job_description:
            raise ValueError("At least one of job_id or job_description must be provided")
        return self


class InterviewSetupResponse(BaseModel):
    prep_id: uuid.UUID
    status: str


class InterviewPrepStatusResponse(BaseModel):
    prep_id: uuid.UUID
    status: str
    question_count: int = 0


class InterviewChatRequest(BaseModel):
    prep_id: uuid.UUID
    session_id: uuid.UUID | None = None
    answer: str = Field(default="", max_length=5000)


class ChatEvaluation(BaseModel):
    score: float
    strengths: list[str]
    improvements: list[str]
    coaching_tip: str
    sample_answer: str


class ChatQuestion(BaseModel):
    question: str
    category: str
    difficulty: int
    follow_up_hints: list[str] = Field(default_factory=list)


class InterviewChatResponse(BaseModel):
    session_id: uuid.UUID
    evaluation: ChatEvaluation | None = None
    next_question: ChatQuestion | None = None
    session_complete: bool = False
    difficulty: int = 1
    overall_feedback: str | None = None


class InterviewPrepListItem(BaseModel):
    id: uuid.UUID
    job_id: uuid.UUID | None = None
    job_title: str | None = None
    company_name: str | None = None
    status: str
    question_count: int = 0
    created_at: datetime


class InterviewPrepListResponse(BaseModel):
    preps: list[InterviewPrepListItem]
    total: int


class SessionMessage(BaseModel):
    role: str
    content: str
    score: float | None = None
    category: str | None = None
    difficulty: int | None = None


class InterviewSessionDetail(BaseModel):
    id: uuid.UUID
    prep_id: uuid.UUID
    status: str
    current_difficulty: int
    questions_answered: int
    overall_score: float | None = None
    messages: list[SessionMessage] = Field(default_factory=list)
    overall_feedback: str | None = None
    completed_at: datetime | None = None
    created_at: datetime


class InterviewSessionListItem(BaseModel):
    id: uuid.UUID
    prep_id: uuid.UUID
    job_title: str | None = None
    company_name: str | None = None
    status: str
    questions_answered: int
    overall_score: float | None = None
    created_at: datetime


class InterviewSessionListResponse(BaseModel):
    sessions: list[InterviewSessionListItem]
    total: int
